File: packages/ios-im-ui/Sources/FlareIMUI/Platform/FlarePlatform.swift
import Foundation
import SwiftUI

// Layer 5 — Platform Contract (shared truth: spec/platform-contract.json).
// One capability record, one adapter interface and one error model. The host
// declares what it can do and performs the native work (PhotosPicker,
// fileImporter, ShareLink …); components read capabilities through
// `@Environment(\.flarePlatform)` and never branch on platform identity.

public enum FlarePlatformKind: String, Sendable { case web, tauri, ios, android, flutter }

public enum FlareCapabilitySupport: String, Sendable { case supported, fallback, unsupported }

public enum FlarePointerKind: String, Sendable { case fine, coarse, mixed, unknown }

public enum FlarePlatformErrorCode: String, Sendable {
    case unsupported = "UNSUPPORTED"
    case cancelled = "CANCELLED"
    case permissionDenied = "PERMISSION_DENIED"
    case timeout = "TIMEOUT"
    case failed = "FAILED"
}

public struct FlarePlatformCapabilities: Equatable, Sendable {
    public var pointer: FlarePointerKind
    public var hover: Bool
    public var contextMenu: Bool
    public var keyboardShortcut: Bool
    /// Contextual layers present as bottom sheets (phone form factor).
    public var bottomSheet: Bool
    public var nativeBack: Bool
    public var safeArea: Bool
    public var filePicker: FlareCapabilitySupport
    public var imagePicker: FlareCapabilitySupport
    public var share: FlareCapabilitySupport

    public init(pointer: FlarePointerKind = .unknown, hover: Bool = false, contextMenu: Bool = false,
                keyboardShortcut: Bool = false, bottomSheet: Bool = false, nativeBack: Bool = false,
                safeArea: Bool = false, filePicker: FlareCapabilitySupport = .unsupported,
                imagePicker: FlareCapabilitySupport = .unsupported, share: FlareCapabilitySupport = .unsupported) {
        self.pointer = pointer; self.hover = hover; self.contextMenu = contextMenu
        self.keyboardShortcut = keyboardShortcut; self.bottomSheet = bottomSheet
        self.nativeBack = nativeBack; self.safeArea = safeArea
        self.filePicker = filePicker; self.imagePicker = imagePicker; self.share = share
    }

    /// What an iPhone / iPad determines on its own: coarse pointer, sheet-first
    /// below the compact width, safe area always; a hardware pointer (iPad
    /// trackpad) turns hover / context menus on. Pickers stay unsupported until
    /// the host adapter declares them.
    public static func ios(width: CGFloat = .infinity, hasPointer: Bool = false) -> FlarePlatformCapabilities {
        FlarePlatformCapabilities(
            pointer: hasPointer ? .mixed : .coarse,
            hover: hasPointer, contextMenu: hasPointer, keyboardShortcut: hasPointer,
            bottomSheet: width < 600, nativeBack: false, safeArea: true
        )
    }
}

/// The contract's error model; `cause` keeps the original error.
public struct FlarePlatformError: Error, Equatable, CustomStringConvertible {
    public let code: FlarePlatformErrorCode
    public let message: String?
    public let cause: Error?

    public init(_ code: FlarePlatformErrorCode, message: String? = nil, cause: Error? = nil) {
        self.code = code; self.message = message; self.cause = cause
    }

    public static func == (lhs: FlarePlatformError, rhs: FlarePlatformError) -> Bool {
        lhs.code == rhs.code && lhs.message == rhs.message
    }

    public var description: String { "FlarePlatformError(\(code.rawValue)\(message.map { ": \($0)" } ?? ""))" }
}

/// `.success(value)` or `.failure(error)` — the only two shapes an adapter operation returns.
public typealias FlarePlatformResult<T> = Result<T, FlarePlatformError>

public extension Result where Failure == FlarePlatformError {
    var isOk: Bool { if case .success = self { return true } else { return false } }
    var valueOrNil: Success? { try? get() }
    var errorOrNil: FlarePlatformError? { if case let .failure(error) = self { return error } else { return nil } }
    var code: FlarePlatformErrorCode? { errorOrNil?.code }
    static func unsupported(_ operation: String) -> Self {
        .failure(FlarePlatformError(.unsupported, message: "\(operation) is not provided by this host"))
    }
}

public struct FlarePickedFile: Equatable, Sendable {
    public let name: String
    public let size: Int?
    public let mimeType: String?
    /// Native filesystem path when the platform exposes one.
    public let path: String?
    /// A URL / uri string when the platform exposes one instead of a path.
    public let uri: String?
    public init(name: String, size: Int? = nil, mimeType: String? = nil, path: String? = nil, uri: String? = nil) {
        self.name = name; self.size = size; self.mimeType = mimeType; self.path = path; self.uri = uri
    }
}

public struct FlarePickFilesOptions: Sendable {
    public let multiple: Bool
    /// MIME types or extensions, e.g. ["image/*", ".pdf"].
    public let accept: [String]
    public init(multiple: Bool = false, accept: [String] = []) { self.multiple = multiple; self.accept = accept }
}

public struct FlarePickImagesOptions: Sendable {
    public let multiple: Bool
    public let video: Bool
    public init(multiple: Bool = false, video: Bool = false) { self.multiple = multiple; self.video = video }
}

public struct FlareSharePayload: Sendable {
    public let title: String?
    public let text: String?
    public let url: String?
    public let files: [FlarePickedFile]
    public init(title: String? = nil, text: String? = nil, url: String? = nil, files: [FlarePickedFile] = []) {
        self.title = title; self.text = text; self.url = url; self.files = files
    }
}

public struct FlareSafeAreaInsets: Equatable, Sendable {
    public let top: CGFloat, right: CGFloat, bottom: CGFloat, left: CGFloat
    public init(top: CGFloat = 0, right: CGFloat = 0, bottom: CGFloat = 0, left: CGFloat = 0) {
        self.top = top; self.right = right; self.bottom = bottom; self.left = left
    }
}

/// Host-implemented native operations. Every operation defaults to
/// UNSUPPORTED, so a host implements only what it can do and declares it in
/// `capabilities`.
public protocol FlarePlatformAdapter {
    var kind: FlarePlatformKind { get }
    var capabilities: FlarePlatformCapabilities { get }
    func pickFiles(_ options: FlarePickFilesOptions) async -> FlarePlatformResult<[FlarePickedFile]>
    func pickImages(_ options: FlarePickImagesOptions) async -> FlarePlatformResult<[FlarePickedFile]>
    func share(_ payload: FlareSharePayload) async -> FlarePlatformResult<Void>
    /// Returns the unsubscribe, or nil when the host has no native back to intercept.
    func onNativeBack(_ handler: @escaping () -> Bool) -> (() -> Void)?
    func safeAreaInsets() -> FlareSafeAreaInsets
}

public extension FlarePlatformAdapter {
    var kind: FlarePlatformKind { .ios }
    func pickFiles(_ options: FlarePickFilesOptions) async -> FlarePlatformResult<[FlarePickedFile]> { .unsupported("pickFiles") }
    func pickImages(_ options: FlarePickImagesOptions) async -> FlarePlatformResult<[FlarePickedFile]> { .unsupported("pickImages") }
    func share(_ payload: FlareSharePayload) async -> FlarePlatformResult<Void> { .unsupported("share") }
    func onNativeBack(_ handler: @escaping () -> Bool) -> (() -> Void)? { nil }
    func safeAreaInsets() -> FlareSafeAreaInsets { FlareSafeAreaInsets() }
}

/// The adapter in effect when the host installs none: nothing native is available.
public struct FlareUnsupportedPlatformAdapter: FlarePlatformAdapter {
    public let capabilities: FlarePlatformCapabilities
    public init(capabilities: FlarePlatformCapabilities = .ios()) { self.capabilities = capabilities }
}

private func matches(_ pattern: String, _ text: String) -> Bool {
    text.range(of: pattern, options: [.regularExpression, .caseInsensitive]) != nil
}

/// Map a thrown error to the contract's error model: `CancellationError` /
/// `NSUserCancelledError` / `NSURLErrorCancelled` → CANCELLED, `NSURLErrorTimedOut`
/// → TIMEOUT, Cocoa no-permission codes → PERMISSION_DENIED,
/// `NSFeatureUnsupportedError` → UNSUPPORTED, anything else FAILED (message
/// heuristics as a last resort).
public func normalizeFlarePlatformError(_ error: Error) -> FlarePlatformError {
    if let platform = error as? FlarePlatformError { return platform }
    let nsError = error as NSError
    let message = nsError.localizedDescription
    func result(_ code: FlarePlatformErrorCode) -> FlarePlatformError { FlarePlatformError(code, message: message, cause: error) }
    if error is CancellationError { return result(.cancelled) }
    switch (nsError.domain, nsError.code) {
    case (NSCocoaErrorDomain, NSUserCancelledError), (NSURLErrorDomain, NSURLErrorCancelled):
        return result(.cancelled)
    case (NSURLErrorDomain, NSURLErrorTimedOut):
        return result(.timeout)
    case (NSCocoaErrorDomain, NSFileReadNoPermissionError), (NSCocoaErrorDomain, NSFileWriteNoPermissionError):
        return result(.permissionDenied)
    case (NSCocoaErrorDomain, NSFeatureUnsupportedError):
        return result(.unsupported)
    default:
        break
    }
    let text = "\(nsError.domain) \(message)"
    if matches("time(d)? ?out", text) { return result(.timeout) }
    if matches("cancel(l)?ed|abort", text) { return result(.cancelled) }
    if matches("permission|denied|not allowed", text) { return result(.permissionDenied) }
    if matches("unsupported|not supported|not implemented", text) { return result(.unsupported) }
    return result(.failed)
}

/// Resolve to TIMEOUT when the native surface does not answer within `milliseconds`.
public func withFlarePlatformTimeout<T: Sendable>(
    milliseconds: UInt64,
    _ operation: @escaping @Sendable () async -> FlarePlatformResult<T>
) async -> FlarePlatformResult<T> {
    await withTaskGroup(of: FlarePlatformResult<T>.self) { group in
        group.addTask { await operation() }
        group.addTask {
            try? await Task.sleep(nanoseconds: milliseconds * 1_000_000)
            return .failure(FlarePlatformError(.timeout, message: "platform operation exceeded \(milliseconds)ms"))
        }
        let first = await group.next() ?? .failure(FlarePlatformError(.failed, message: "no result"))
        group.cancelAll()
        return first
    }
}

/// Run an adapter operation through the contract: a thrown error is normalized
/// and an optional timeout applies.
public func callFlarePlatform<T: Sendable>(
    timeoutMilliseconds: UInt64? = nil,
    _ operation: @escaping @Sendable () async throws -> FlarePlatformResult<T>
) async -> FlarePlatformResult<T> {
    let guarded: @Sendable () async -> FlarePlatformResult<T> = {
        do { return try await operation() } catch { return .failure(normalizeFlarePlatformError(error)) }
    }
    if let timeoutMilliseconds { return await withFlarePlatformTimeout(milliseconds: timeoutMilliseconds, guarded) }
    return await guarded()
}

private struct FlarePlatformKey: EnvironmentKey {
    static let defaultValue: any FlarePlatformAdapter = FlareUnsupportedPlatformAdapter()
}

public extension EnvironmentValues {
    /// The host adapter in effect (the unsupported fallback when none is installed).
    var flarePlatform: any FlarePlatformAdapter {
        get { self[FlarePlatformKey.self] }
        set { self[FlarePlatformKey.self] = newValue }
    }
}

public extension View {
    /// Installs the host's platform adapter for this subtree. Call once near the app root.
    func flarePlatform(_ adapter: any FlarePlatformAdapter) -> some View {
        environment(\.flarePlatform, adapter)
    }
}

File: packages/android-im-ui/src/main/kotlin/com/flare/im/ui/FlarePlatform.kt
package com.flare.im.ui

import androidx.annotation.RequiresApi
import android.content.ActivityNotFoundException
import android.os.Build
import android.provider.Settings
import android.view.WindowManager
import androidx.activity.compose.BackHandler
import androidx.activity.compose.LocalOnBackPressedDispatcherOwner
import androidx.compose.runtime.Composable
import androidx.compose.runtime.CompositionLocalProvider
import androidx.compose.runtime.DisposableEffect
import androidx.compose.runtime.ReadOnlyComposable
import androidx.compose.runtime.remember
import androidx.compose.runtime.staticCompositionLocalOf
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.platform.LocalView
import androidx.compose.ui.window.DialogWindowProvider
import kotlinx.coroutines.TimeoutCancellationException
import kotlinx.coroutines.withTimeout
import kotlin.coroutines.cancellation.CancellationException

// Layer 5 — Platform Contract (shared truth: spec/platform-contract.json).
// One capability record, one adapter interface and one error model. The host
// declares what it can do and performs the native work (ActivityResult
// contracts, share intents …); components read capabilities through
// [flarePlatform] and never branch on platform identity.

enum class FlarePlatformKind { Web, Tauri, Ios, Android, Flutter }

enum class FlareCapabilitySupport { Supported, Fallback, Unsupported }

enum class FlarePointerKind { Fine, Coarse, Mixed, Unknown }

enum class FlarePlatformErrorCode { UNSUPPORTED, CANCELLED, PERMISSION_DENIED, TIMEOUT, FAILED }

data class FlarePlatformCapabilities(
    val pointer: FlarePointerKind = FlarePointerKind.Unknown,
    val hover: Boolean = false,
    val contextMenu: Boolean = false,
    val keyboardShortcut: Boolean = false,
    /** Contextual layers present as bottom sheets (phone form factor). */
    val bottomSheet: Boolean = false,
    val nativeBack: Boolean = false,
    val safeArea: Boolean = false,
    val filePicker: FlareCapabilitySupport = FlareCapabilitySupport.Unsupported,
    val imagePicker: FlareCapabilitySupport = FlareCapabilitySupport.Unsupported,
    val share: FlareCapabilitySupport = FlareCapabilitySupport.Unsupported,
) {
    companion object {
        /** What an Android phone / tablet determines on its own; pickers stay unsupported until the host adapter declares them. */
        fun android(widthDp: Int = Int.MAX_VALUE, hasPointer: Boolean = false): FlarePlatformCapabilities = FlarePlatformCapabilities(
            pointer = if (hasPointer) FlarePointerKind.Mixed else FlarePointerKind.Coarse,
            hover = hasPointer,
            contextMenu = hasPointer,
            keyboardShortcut = hasPointer,
            bottomSheet = widthDp < 600,
            nativeBack = true,
            safeArea = true,
        )
    }
}

/** The contract's error model; [cause] keeps the original throwable. */
class FlarePlatformError(
    val code: FlarePlatformErrorCode,
    message: String? = null,
    cause: Throwable? = null,
) : Exception(message, cause) {
    override fun toString(): String = "FlarePlatformError($code${message?.let { ": $it" } ?: ""})"
}

/** `Ok(value)` or `Err(error)` — the only two shapes an adapter operation returns. */
sealed class FlarePlatformResult<out T> {
    data class Ok<T>(val value: T) : FlarePlatformResult<T>()
    data class Err(val error: FlarePlatformError) : FlarePlatformResult<Nothing>()

    val isOk: Boolean get() = this is Ok
    val valueOrNull: T? get() = (this as? Ok<T>)?.value
    val errorOrNull: FlarePlatformError? get() = (this as? Err)?.error
    val code: FlarePlatformErrorCode? get() = errorOrNull?.code

    companion object {
        fun failure(code: FlarePlatformErrorCode, message: String? = null, cause: Throwable? = null): Err =
            Err(FlarePlatformError(code, message, cause))
    }
}

data class FlarePickedFile(
    val name: String,
    val size: Long? = null,
    val mimeType: String? = null,
    /** Native filesystem path when the platform exposes one. */
    val path: String? = null,
    /** Content uri when the platform exposes one instead of a path. */
    val uri: String? = null,
)

data class FlarePickFilesOptions(val multiple: Boolean = false, val accept: List<String> = emptyList())

data class FlarePickImagesOptions(val multiple: Boolean = false, val video: Boolean = false)

data class FlareSharePayload(
    val title: String? = null,
    val text: String? = null,
    val url: String? = null,
    val files: List<FlarePickedFile> = emptyList(),
)

data class FlareSafeAreaInsets(val top: Float = 0f, val right: Float = 0f, val bottom: Float = 0f, val left: Float = 0f)

/**
 * Host-implemented native operations. Every operation defaults to UNSUPPORTED,
 * so a host overrides only what it can do and declares it in [capabilities].
 */
interface FlarePlatformAdapter {
    val kind: FlarePlatformKind get() = FlarePlatformKind.Android
    val capabilities: FlarePlatformCapabilities

    suspend fun pickFiles(options: FlarePickFilesOptions = FlarePickFilesOptions()): FlarePlatformResult<List<FlarePickedFile>> =
        FlarePlatformResult.failure(FlarePlatformErrorCode.UNSUPPORTED, "pickFiles is not provided by this host")

    suspend fun pickImages(options: FlarePickImagesOptions = FlarePickImagesOptions()): FlarePlatformResult<List<FlarePickedFile>> =
        FlarePlatformResult.failure(FlarePlatformErrorCode.UNSUPPORTED, "pickImages is not provided by this host")

    suspend fun share(payload: FlareSharePayload): FlarePlatformResult<Unit> =
        FlarePlatformResult.failure(FlarePlatformErrorCode.UNSUPPORTED, "share is not provided by this host")

    /** Returns the unsubscribe, or null when the host has no native back to intercept. */
    fun onNativeBack(handler: () -> Boolean): (() -> Unit)? = null

    fun safeAreaInsets(): FlareSafeAreaInsets = FlareSafeAreaInsets()
}

/** The adapter in effect when the host installs none: nothing native is available. */
class FlareUnsupportedPlatformAdapter(
    override val capabilities: FlarePlatformCapabilities = FlarePlatformCapabilities.android(),
) : FlarePlatformAdapter

private val cancelPattern = Regex("cancel(l)?ed|abort", RegexOption.IGNORE_CASE)
private val permissionPattern = Regex("permission|denied|not allowed", RegexOption.IGNORE_CASE)
private val timeoutPattern = Regex("time(d)? ?out", RegexOption.IGNORE_CASE)
private val unsupportedPattern = Regex("unsupported|not supported|not implemented", RegexOption.IGNORE_CASE)

/**
 * Map a thrown value to the contract's error model: [TimeoutCancellationException]
 * → TIMEOUT, [CancellationException] → CANCELLED, [SecurityException] →
 * PERMISSION_DENIED, [ActivityNotFoundException] / [UnsupportedOperationException]
 * → UNSUPPORTED, anything else FAILED (message heuristics as a last resort).
 */
fun normalizeFlarePlatformError(error: Throwable): FlarePlatformError {
    if (error is FlarePlatformError) return error
    // Android framework exceptions built in JVM unit tests may carry no message; the contract always keeps one.
    val message = error.message ?: error::class.simpleName ?: error.javaClass.name
    val text = "${error::class.simpleName} $message"
    fun result(code: FlarePlatformErrorCode) = FlarePlatformError(code, message, error)
    return when {
        error is TimeoutCancellationException || timeoutPattern.containsMatchIn(text) -> result(FlarePlatformErrorCode.TIMEOUT)
        error is CancellationException || cancelPattern.containsMatchIn(text) -> result(FlarePlatformErrorCode.CANCELLED)
        error is SecurityException || permissionPattern.containsMatchIn(text) -> result(FlarePlatformErrorCode.PERMISSION_DENIED)
        error is ActivityNotFoundException || error is UnsupportedOperationException || error is NotImplementedError ||
            unsupportedPattern.containsMatchIn(text) -> result(FlarePlatformErrorCode.UNSUPPORTED)
        else -> result(FlarePlatformErrorCode.FAILED)
    }
}

/** Resolve to TIMEOUT when the native surface does not answer within [timeoutMs]. */
suspend fun <T> withFlarePlatformTimeout(timeoutMs: Long, operation: suspend () -> FlarePlatformResult<T>): FlarePlatformResult<T> =
    try {
        withTimeout(timeoutMs) { operation() }
    } catch (e: TimeoutCancellationException) {
        FlarePlatformResult.failure(FlarePlatformErrorCode.TIMEOUT, "platform operation exceeded ${timeoutMs}ms", e)
    }

/** Run an adapter operation through the contract: a thrown value is normalized and an optional timeout applies. */
suspend fun <T> callFlarePlatform(timeoutMs: Long? = null, operation: suspend () -> FlarePlatformResult<T>): FlarePlatformResult<T> =
    try {
        if (timeoutMs == null) operation() else withFlarePlatformTimeout(timeoutMs, operation)
    } catch (e: Throwable) {
        FlarePlatformResult.Err(normalizeFlarePlatformError(e))
    }

val LocalFlarePlatform = staticCompositionLocalOf<FlarePlatformAdapter> { FlareUnsupportedPlatformAdapter() }

/** Installs the host's [FlarePlatformAdapter] for a subtree; read it with [flarePlatform]. */
@Composable
fun FlarePlatformProvider(adapter: FlarePlatformAdapter, content: @Composable () -> Unit) {
    CompositionLocalProvider(LocalFlarePlatform provides adapter, content = content)
}

/** The host adapter in effect (the unsupported fallback when none is installed). */
@Composable
@ReadOnlyComposable
fun flarePlatform(): FlarePlatformAdapter = LocalFlarePlatform.current

/**
 * System back for a surface that shows a back control: while [enabled], the OS back
 * gesture / button runs [onBack] instead of leaving the host. Registered only when the
 * host declares `nativeBack` and provides an OnBackPressedDispatcher (any
 * ComponentActivity does); the innermost enabled surface wins.
 */
@Composable
internal fun FlareNativeBackEffect(enabled: Boolean, onBack: () -> Unit) {
    if (!flarePlatform().capabilities.nativeBack || LocalOnBackPressedDispatcherOwner.current == null) return
    BackHandler(enabled = enabled, onBack = onBack)
}

/**
 * Reduced motion: the user removed animations (animator duration scale 0). Kit motion that
 * moves content — sheet slides, list scrolls — then jumps straight to its end state.
 */
@Composable
internal fun flareReducedMotion(): Boolean {
    val resolver = LocalContext.current.contentResolver
    return remember(resolver) { Settings.Global.getFloat(resolver, Settings.Global.ANIMATOR_DURATION_SCALE, 1f) == 0f }
}

/**
 * Makes the dialog window this is composed in take the whole display. A full-screen dialog window
 * must not be pushed below the status bar (its bottom would fall off screen): it lays out under the
 * bars and the display cutout, and its content applies the insets it needs. Reads the API level, so it
 * lives with the platform module.
 */
@Composable
internal fun FlareEdgeToEdgeDialogWindow() {
    val window = (LocalView.current.parent as? DialogWindowProvider)?.window
    DisposableEffect(window) {
        if (window != null && Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
            window.attributes = window.attributes.apply {
                layoutInDisplayCutoutMode = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
                    WindowManager.LayoutParams.LAYOUT_IN_DISPLAY_CUTOUT_MODE_ALWAYS
                } else {
                    WindowManager.LayoutParams.LAYOUT_IN_DISPLAY_CUTOUT_MODE_SHORT_EDGES
                }
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) fitInsetsTypes = 0
            }
        }
        onDispose {}
    }
}

